File: src/main/resources/scripts/Plugins/PDCSL/Nuclei_Counter.py
# NucleiCounter - Count cell nuclei
# Copyright © 2019 Damien Goutte-Gattat

#@ Boolean (label="Process entire stack", default=true) do_stack

from ij import IJ, ImagePlus
from ij.measure import ResultsTable
from ij.plugin.filter import ParticleAnalyzer

from org.incenp.imagej import NucleiSegmenter


def run_script(do_stack):
    image = IJ.getImage()
    segmenter = NucleiSegmenter(2.0)
    results = ResultsTable()
    flags = ParticleAnalyzer.DISPLAY_SUMMARY
    if not do_stack:
        flags |= ParticleAnalyzer.SHOW_OUTLINES
    analyzer = ParticleAnalyzer(flags, 0, results, 0.0, 500.0, 0.0, 1.0)

    if do_stack:
        segmented = segmenter.segment(image, [i + 1 for i in range(image.getNChannels())])
        counter = 0
        values = [0 for n in range(segmented.getNChannels())]

        for i in range(segmented.getNChannels()):
            for j in range(segmented.getNSlices()):
                segmented.setPosition(i + 1, j + 1, 1)

                if analyzer.analyze(segmented):
                    n = results.getCounter() - counter
                    counter = results.getCounter()
                    values[i] += n

        fmt = "{}" + ",{}" * len(values)
        IJ.log(fmt.format(image.getTitle(), *values))
    else:
        segmented = ImagePlus("binary", segmenter.segment(image.getProcessor()))
        segmented.show()
        if analyzer.analyze(segmented):
            output = analyzer.getOutputImage()
            output.show()


run_script(do_stack)

File: src/main/java/uk/ac/qmul/bci/pdcsl/imagej/NucleiCounter.java
/*
 * PDCSL-ImageJ - PDCS Lab’s ImageJ Collection
 * Copyright © 2019, 2021 Damien Goutte-Gattat
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the Gnu General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package uk.ac.qmul.bci.pdcsl.imagej;

import org.incenp.imagej.Helper;
import org.incenp.imagej.NucleiSegmenter;
import org.scijava.command.Command;
import org.scijava.plugin.Parameter;
import org.scijava.plugin.Plugin;
import org.scijava.ui.UIService;

import ij.ImagePlus;
import ij.measure.ResultsTable;

@Plugin(type = Command.class, menuPath = "Plugins>PDCSL>Nuclei Counter")
public class NucleiCounter implements Command {

    @Parameter
    private ImagePlus image;

    @Parameter(label = "Channel(s) to process", required = false)
    private String channelList;

    @Parameter(label = "Show the segmented image")
    private boolean showSegmentedImage;

    @Parameter
    private UIService uiService;

    @Override
    public void run() {
        NucleiSegmenter ns = new NucleiSegmenter(2.0);
        int[] channels = Util.parseChannels(image, channelList);

        ImagePlus segmented = ns.segment(image, channels);
        int[] results = Util.countParticles(segmented, 0, 500.0, channels);

        ResultsTable rt = Helper.getResultsTable("Nuclei Counter");
        rt.incrementCounter();
        rt.addLabel(image.getTitle());
        for ( int i = 0; i < channels.length; i++ )
            rt.addValue(String.format("Channel %d", i + 1), results[i]);
        rt.show("Nuclei Counter");

        if ( showSegmentedImage )
            uiService.show(segmented);
        else
            segmented.close();
    }

}
